File: app/src/main/java/com/mc10inc/codetest/login/ErrorDialogFragment.java
package com.mc10inc.codetest.login;

import android.app.Dialog;
import android.content.DialogInterface;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v4.app.DialogFragment;
import android.support.v4.app.FragmentManager;
import android.support.v7.app.AlertDialog;

import com.mc10inc.codetest.R;

import io.reactivex.Observable;
import io.reactivex.subjects.PublishSubject;

/**
 * TODO
 *
 * @author Adam Stroud &#60;<a href="mailto:adam.stroud@gmail.com">adam.stroud@gmail.com</a>&#62;
 */
public class ErrorDialogFragment extends DialogFragment {
    private final PublishSubject<ClickEvent> clickSubject = PublishSubject.create();

    @NonNull
    @Override
    public Dialog onCreateDialog(Bundle savedInstanceState) {
        return new AlertDialog.Builder(getContext())
                .setMessage(R.string.message_dialog_login_error)
                .setCancelable(false)
                .setPositiveButton(android.R.string.ok, (dialog, which) -> clickSubject.onNext(new ClickEvent(dialog, which)))
                .setOnDismissListener(dialog -> clickSubject.onComplete())
                .create();
    }

    public Observable<ClickEvent> show(FragmentManager manager)  {
        super.show(manager, ErrorDialogFragment.class.getSimpleName());
        return clickSubject;
    }

    public static class ClickEvent {
        private final DialogInterface dialog;
        private final int which;

        public ClickEvent(DialogInterface dialog, int which) {
            this.dialog = dialog;
            this.which = which;
        }

        public DialogInterface getDialog() {
            return dialog;
        }

        public int getWhich() {
            return which;
        }
    }
}

File: app/src/main/kotlin/com/mc10inc/codetest/login/ErrorDialogFragment.kt
package com.mc10inc.codetest.login

import android.app.Dialog
import android.content.DialogInterface
import android.os.Bundle
import android.support.v4.app.DialogFragment
import android.support.v4.app.FragmentManager
import android.support.v7.app.AlertDialog
import com.mc10inc.codetest.R
import io.reactivex.Observable
import io.reactivex.subjects.PublishSubject

/**
 * TODO
 *
 * @author Adam Stroud &#60;[adam.stroud@gmail.com](mailto:adam.stroud@gmail.com)&#62;
 */
class ErrorDialogFragment : DialogFragment() {
    private val clickSubject = PublishSubject.create<ClickEvent>()

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        return AlertDialog.Builder(context!!)
                .setMessage(R.string.message_dialog_login_error)
                .setCancelable(false)
                .setPositiveButton(android.R.string.ok, { dialog, which ->
                    clickSubject.onNext(ClickEvent(dialog, which)) })
                .setOnDismissListener { clickSubject.onComplete() }
                .create()
    }

    fun show(manager: FragmentManager?): Observable<ClickEvent> {
        super.show(manager, ErrorDialogFragment::class.java.simpleName)
        return clickSubject
    }

    inner class ClickEvent(val dialog: DialogInterface, val which: Int)
}